docs/spec/identities.md

```diff
@@ -0,0 +1,438 @@
+
+    * The document does not refer to a previous revision, and no previous
+      revision is known
+    * **Or**, the set of signatures forms a quorum of the delegations of the
+      **previous** revision.
+
+    It is an error if:
+
+    * No previous revision is given, but a parent in the identity attestation
+      chain is found
+    * A previous revision is given, but the identity attestation chain does not
+      yield a parent
+    * A previous revision is given, but it is not the same the parent
+      attestation refers to
+    * The current and parent attestations refer to different `root`s
+
+The verification process can now be described recursively:
+
+```rust
+/* State transitions, definitions elided for brevity */
+fn signed(i: Untrusted<Identity>) -> Result<Signed<Identity>, Error>;
+fn quorum(i: Signed<Identity>) -> Result<Quorum<Identity>, Error>;
+fn verified(i: Quorum<Identity>, parent: Option<Verified<Identity>>)
+    -> Result<Verified<Identity>, Error>;
+
+fn verify(
+    head: Untrusted<Identity>,
+    mut parents: impl Iterator<Item = Untrusted<Identity>>,
+) -> Result<Verified<Identity>, Error> {
+    let head = quorum(signed(head)?)?;
+    let parent = match parents.next() {
+        None => Ok(None),
+        Some(parent) => verify(parent, parents),
+    }?;
+
+    verified(head, parent)
+}
+```
+
+> **Implementation Notes**:
+>
+> * The recusive definition is given for brevity, implementations may prefer to
+>   walk the history in reverse order.
+> * As an optimisation, implementations SHOULD store the result of a successful
+>   verification in persistent storage, such that verification of updates can
+>   start from the last-good state instead of the history root. Observe,
+>   however, that two different `Identity` objects (as identitified by their
+>   `id`) may attest the same `Doc` revision -- the persistent state SHOULD
+>   therefore allow retrieving the set of attesting `Identity`s by `Doc`
+>   revision.
+> * The `quorum` predicate above SHOULD actually skip over this parent instead
+>   of aborting the procedure -- it is possible that a later `Identity` reaches
+>   the quorum.
+
+## Delegations
+
+Radicle Link distinguishes two types of identities: personal and project. The
+first describes an actor in the system, while the second describes a (software)
+project on which one or more actors collaborate. Apart from their payload types
+`T`, they differ in their delegations type `D`:
+
+Personal identities can only delegate to anonymous keys, while project
+identities MAY attach a personal identity (by it's URN) to a key delegation.
+
+More formally:
+
+```rust
+type User<T> = Identity<T, HashSet<PublicKey>>;
+type Project<T> = Identity<T, HashMap<PublicKey, Option<Urn>>>;
+```
+
+> **TODO** Consider a container for project delegations which does not require
+> to repeat the URN for every key
+
+The set of `PublicKey`s delegated to MUST be unique per identity document.
```

Comment:
How can one produce a new revision of an entity with different metadata (the T payload) if one does not want to change the set of delegated keys?
This should be a common operation but it seems forbidden by this rule.

Reply:
Why should this be a common operation? The **whole point** of this is to secure who can modify the metadata. If you want to separate those two, you're designing a PKI.

Reply:
> Modifying level 2 (like, changing the user email or name in the profile) should not require changing the set of keys at level 1.

It is completely irrelevant whether or not the keys change when the metadata changes. The point is that it is those keys which are allowed to change it, and so a new revision must be created in any case. What is the problem you're trying to solve?

Reply:
I was trying to understand this sentence:

```
The set of `PublicKey`s delegated to MUST be unique per identity document.
```

Does it mean that there cannot be two entity revisions sharing the same set of keys?
Or did I parse it incorrectly, and "identity document" means the stable URN? (the 1st revision of the feed of entity documents)

BTW, two different projects *could* delegate to the exact set of maintainers, therefore to the exact set of keys.
This is what I am not getting.

Anyway, this explanation:

> It is completely irrelevant whether or not the keys change when the metadata changes.

clarifies the point for me, and on this we agree.
It is that spec line that I am not understanding.

Reply:
> The set of `PublicKey`s delegated to MUST be unique per identity document.

I think all this is saying is that we can't have duplicated `PublicKey`s in the set of keys. This is clear when we read the pseudo-code `HashSet<PublicKey>`, but the RFC should apply to any language implementation.

Reply:
> I think all this is saying is that we can't have duplicated `PublicKey`s in the set of keys.

Ah, that's ok.

But the wording is really misleading for me.
I mean, I cannot imagine a "set" with not unique elements, regardless of the programming language.
So I parsed it as "the whole set must be unique among different documents", and could not understand that at all :-/

Could we word it like this?
```
The set of `PublicKey`s delegated to in an identity MUST not contain duplicates.
```